### DISCUSSION TITLE
Поговорим о Wiring

### DISCUSSION THREAD

**Discussion by AnatolSher** [board707/w80x_arduino#11] [General]
Дмитрий, а у Вас есть понимание или представление как проинициализировать борду? Так как китайцы сделали, это отстой.
В моем понимании, сначала нужно собрать все данные о том какую периферию хочет задействовать ардуино-прогер. Затолкать все параметры в какую либо структуру однозначно описывающее конкретную борду, и сделать вызов функции инициализации железа один раз, непосредственно перед передачей управления в loop(). И больше не проверять.
Китайцы же проверяют соответствие пинов и режимов постоянно, постоянно ремапят... и т.д В общем не понятно для чего эти костыли.

**Comment by board707** [comment]
Не могу сказать, что у меня есть четкое понимание, но судя по другим платам, в пакетах ардуино принято инициализировать всю встроенную периферию по максимуму заранее. Пользователь не должен в явном виде запускать какую-то инициализацию, он пишет только свой пользовательский код.

Насчет того, что китайцы постоянно что-то ремапят - не могли бы Вы привести пример? Я если это речь про наш пакет - там ардуино-кода-то пяток строчек, когда они успевают? :)

**Reply by AnatolSher** [reply]
В дебаггере видно. Достаточно протрассировать analogWrite() или analogRead()

**Comment by board707** [comment]
Многие знания - многие печали... :)

Если вы имеете в виду перепроверку и переинициализацию пинов - это в Ардуино во всех платах так. При каждом обращении пин инициализируется заново.

Добавлю - для ардуино мира это абсолютно нормально. Если вы мигаете лампочкой с частотой раз в секунду, вам абсолютно все равно, занимает ли вывод на пин 1 микросекунду или даже пару миллисекунд.
Те, для кого эта разница важна - просто не используют digitalWrite() или analogRead()

**Reply by AnatolSher** [reply]
Это не эффективно. Потеря машинного времени. Все проверки по использованию периферии и их непротиворечивости надо делать в самом начале. В КубеМХ например мы визуально видим конфликты при применении альтернатив. Здесь на этапе инициализации борды нужно выкатывать в USART0 все косяки и стопорить процессор

**Reply by AnatolSher** [reply]
Кстати, есть какой либо подход к подбору чисел, которые в настройках PWM дадут целые?
Я сейчас такие параметры использую
![image](https://user-images.githubusercontent.com/121094544/219031943-b9d0b943-2232-44ab-8390-1d948d9ca5d1.png)
В итоге, когда в analogWrite(pin,val) Значение val=0xFF, на осциллограмме остается хвостик. Т.е уровень не принимает стабильного HIGH при 100% заполнении

upd Судя по формуле это недостижимо?

**Comment by board707** [comment]
Мне кажется, Вы мыслите категориями больших компьютеров. В микро мире у пользователя на этапе инициализации консоль на USART0 может и отсуствовать и он никакой вашей диагностики не увидит. Для него это будет выглядеть так, как будто плата зависла.
В ардуино мире код проектируется так, чтобы им могли пользоваться 10-летние дети.

Мне кажется, что о эффективности тут надо думать в последнюю очередь. Еще раз - тем, кому нужна эффективность - не нужны команды ардуино, они будут пользовать функции СДК напрямую.

Посмотрите аддон Эрла Филхофера для ярда RP2040. Он поддерживает функции Вайринга, но вместе с тем он дает прямой доступ ко всем методам СДК. И весь эффективный код, конечно, пишется через СДК, а не через Вайринг.
Если бы нам удалось сделать что-то похожее - я бы считал задачу выполненной

**Reply by AnatolSher** [reply]
Ну и сейчас можно миксировать, wiring + hal style
Другой вопрос... wiring в зачаточном состоянии

**Comment by board707** [comment]
Вообще главный смысл Вайринга - переносимость кода.  Если вы используете только ардуино-функции, один и тот же код будет работать и на Ардуино Нано, и на ЕСП32, и на СТМ.
То есть пользователь может запустить на нашей плате код для Нано и он должен работать.

Если вы начнете требовать от пользователя инициализировать периферию "вручную" - никакой переносимости и близко не останется.

**Reply by AnatolSher** [reply]
> Если вы начнете требовать от пользователя инициализировать периферию "вручную" - никакой переносимости и близко не останется.

Нет, я не про это. Для ардуинщика ничего не меняется. Он как в секции setup() описывает настройки пинов, так и будет продолжать. Все проверки скрыты от его глаз. Я лишь о самой концепции инициализации борды говорю, т.е системный уровень
